### DISCUSSION TITLE
Cannot select/connect bluetooth device

### DISCUSSION THREAD

**Discussion by manio** [cedricp/ecutweaker#19] [Q&A] [answered]
Hi Cedric,
According to the wiki here:
https://github.com/cedricp/ddt4all/wiki/Android-port
I downloaded and installed the latest version there `(DOWNLOAD LINK V1.9) -> EcuTweaker-1.9.apk`
And it seems this particular version has a bug which leads to problem with connecting to bluetooth devices:
when clicking on `Select device` button (in bluetooth mode/icon) nothing happens (bluetooth persmission has been granted).

When I downgraded to previous version: `EcuTweaker-release-v1.85.apk` then all is working just fine.

Maybe this bug was fixed in the mean time (sorry for reporting then), I only want to report this, so you are aware of this problem.

Tested on: Galaxy A10 / Android 11

**Comment by Furtif** [comment]
latest version is https://github.com/cedricp/ecutweaker/releases/tag/v2.0

**Comment by Furtif** [comment]
uses [debug](https://github.com/cedricp/ecutweaker/releases/download/v2.0/EcuTweaker-debug.apk) build

**Comment by Furtif** [comment]
Fine ??? If ok Close this issue.
Thanks.

**Comment by manio** [comment]
I have it on my radar.
Please leave it open for a few days, I'll give it a try with newest version ASAP...

**Comment by manio** [comment]
Hi,
Unfortunately i cannot test the bluetooth because the version 2.0 which you've mentioned is not working for me - it is constantly crashing at startup (`Ecu Tweaker keeps stopping: App info/Close app`) on my Samsung Galaxy A10 phone. Then I've got notification:
`Ecu Tweaker is crashing frequently. Put it in deep sleep to keep your phone running smoothly.`

So for me the version which is working great is still: `EcuTweaker-release-v1.85.apk`

**Comment by Furtif** [comment]
you uses a good DB ecu-zip ?? try this https://github.com/cedricp/ddt4all/discussions/845

**Comment by Furtif** [comment]
place zip in InternalStorage root

**Comment by Furtif** [comment]
can't not unzip place as zip

**Comment by Furtif** [comment]
and not uses Release apk uses the [EcuTweaker-debug.apk](https://github.com/cedricp/ecutweaker/releases/download/v2.0/EcuTweaker-debug.apk) apk release apk is bugued

**Comment by manio** [comment]
Yes, this is the database I am using. It's md5 is exactly:
`57d9856cf2959b387c408867cb32dfdf  ecu.zip`
I know how to use it because the earlier version is working fine (with the same ecu.zip)

I will now try the debug version...

**Comment by manio** [comment]
Ok ... the debug version started fine!
... but the bluetooth bug is still there. I am clicking on `SELECT DEVICE` and nothing happen.

When I change eg for USB, then button is doing action (eg `EcuTweaker USB scan`).

**Comment by Furtif** [comment]
i think issue in release mode is because obfucation mode https://github.com/cedricp/ecutweaker/blob/master/EcuTweaker/build.gradle#L46

for bluethoo maybe you need associate device before in android settings>conexions->bluethoot

**Comment by manio** [comment]
I had the device paired before app start (and it is working without any problems on older versions). The bug probably started between version 1.85 and 1.9.

**Comment by Furtif** [comment]
@manio  try new release https://github.com/cedricp/ecutweaker/releases/tag/v2.1

release or debug normally, both are ok, I tested and the bluethood devices already appear.

**Comment by manio** [comment]
Thanks, I've installed this 2.1 release version and unfortunately no change - clicking on "Select device" -> no action.

**Comment by Furtif** [comment]
odd!!

**Comment by weaydenkim** [comment]
I have the same problem.
There was no problem with using the 1.85 version
All 1.9, 2.0, 2.1v are not working.
I also downloaded a new ECU.ZIP file and tested it,
Unfortunately, it doesn't work.

**Comment by dwazero** [comment]
Hello, versions 2.0 also do not work for me (it turns off because it detects an error), 2.1 release and 2.1 debug, both of them start up, but do not detect the bluetoth device. Versions 1.85 and 1.9 work correctly, the others do not detect bluetoth
on the phone:
Galaxy a53 5g android 14 oneui 6.0
I will add Bluetooth devices, both versions, i.e. 2.1 debug and 2.1 release, are detected but cannot connect

**Comment by cedricp** [comment]
Could you try to  launch the app in dbg mode and read logcat (with android studio). Maybe there are some interesting warnings/errors inside it to fix that issue.

**Comment by dwazero** [comment]
I already installed 1.9

If you can, try to make it so that BT connection works in version 2.1 release or debug version

**Comment by dwazero** [comment]
when can we expect the bT problem to be fixed?

**Comment by cedricp** [comment]
As soon as we get more information. A logcat reading could help us localize the bug.
We can't buy every android devices and check against it.

**Comment by dwazero** [comment]
HOW to read this logcat? what would I have to do?download the release or debug version?

**Comment by manio** [comment]
@cedricp 
Never done this before, so few questions:
- Can I do it without installing the whole `Android Studio`? (I have adb).
- Does the phone needs to be rooted?
- I assume I need to somehow use this: https://developer.android.com/tools/logcat,
so `adb logcat` or `adb shell logcat` ?

**Comment by cedricp** [comment]
I'm used to use logcat inside Android studio, but adb liogcat should work as well.
Capture the log while using the app and search for debug/error message.

**Comment by manio** [comment]
@cedricp 
As this log may contain sensitive data I've sent you this via e-mail. It is from application start to clicking on the "Select device" button.

**Comment by dwazero** [comment]
Does BT work for you yet or not?it's about version 2.1 release or debug

**Comment by cedricp** [comment]
> @cedricp As this log may contain sensitive data I've sent you this via e-mail. It is from application start to clicking on the "Select device" button.

Did not receive it, sorry

**Comment by manio** [comment]
> Did not receive it, sorry

Look into the spam folder. Google doesn't like foreign SMTP servers. Besides the attachment may seems like a virus for its anti-spam AI :)

**Comment by manio** [comment]
@dwazero 
Dzięki za info że CI też nie działa - zawsze to jest dodatkowa istotna informacja. Skupiamy sie aktualnie na rozwiązaniu tego problemu. Wiemy od której wersji przestało to działać i autorzy nad tym pracują a ja im próbuję pomóc... więc luzik :)

**Comment by dwazero** [comment]
To czekam na jakieś info :) 1.9 działa poprawnie mimo,że pisze,ze aplikacja jest na starszą wersje androida.Natomiast 2.0 wyłącza się od razu podczas rozruchu jakiś błąd jest a 2.1 zaś nie znajduje BT

**Comment by Furtif** [comment]
try https://github.com/cedricp/ecutweaker/actions/runs/8198600397#artifacts

**Comment by manio** [comment]
Nope :(
Unfortunately either debug or release doesn't allow selecting bluetooth. I was also trying to disable bluetooth and when I press that `Select device` button Android is asking me whether to turn it on ... so it is some kind working but no window at all.

Maybe git bisect what happens between last good working and first "bad" one?

**Comment by Furtif** [comment]
This is very strange, the permissions are good etc... I have to note one thing.  or add fine location permissions.  I have been reading some Android documentation and they apparently use this for wi-fi/bluetooth

**Comment by dwazero** [comment]
does it already know whether version 2.1 release or debug works correctly and finds BT?

**Comment by Furtif** [comment]
> does it already know whether version 2.1 release or debug works correctly and finds BT?

Not yet, and I don't know why it doesn't work if all permissions are ok. I don't know where this error comes from

**Comment by Furtif** [comment]
let's see if @cedricp has a solution

**Comment by manio** [comment]
Maybe this will help:
|Version 1.85|recent "test" version from 8th march|
|----------|------------|
|working|not working|
|![image](https://github.com/cedricp/ecutweaker/assets/583157/491192fb-5e52-4721-859f-47f2db16581e)|![image](https://github.com/cedricp/ecutweaker/assets/583157/62e26e3c-d5fd-491c-a2ab-2d72acc9ea0a)|

**Comment by Furtif** [comment]
> Maybe this will help:
> 
> Version 1.85	recent "test" version from 8th march
> working	not working
> ![image](https://private-user-images.githubusercontent.com/583157/320575041-491192fb-5e52-4721-859f-47f2db16581e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2MDAyNDEsIm5iZiI6MTcxMjU5OTk0MSwicGF0aCI6Ii81ODMxNTcvMzIwNTc1MDQxLTQ5MTE5MmZiLTVlNTItNDcyMS04NTlmLTQ3ZjJkYjE2NTgxZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwOFQxODEyMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNjM5ODkzZDFiMjkxNGM1NmI4YWMwMGI4NDI5NWFlZGM2MjRlZjg4OTk3MDZlOGI2MDJlZTc2YjZkMDkwNTI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.INfil3kl2uvHZ-jhlnmfta15LT8FwCJt-bmArm8enTk)	![image](https://private-user-images.githubusercontent.com/583157/320575949-62e26e3c-d5fd-491c-a2ab-2d72acc9ea0a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2MDAyNDEsIm5iZiI6MTcxMjU5OTk0MSwicGF0aCI6Ii81ODMxNTcvMzIwNTc1OTQ5LTYyZTI2ZTNjLWQ1ZmQtNDkxYy1hMmFiLTJkNzJhY2M5ZWEwYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwOFQxODEyMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYjc1ZTE1NTgwMjgzM2UwN2ZlMTMwY2Q0NjUzZWI3ZmNiMTJmYjdiNGU5M2Y5MGE3ODNkNDg2NWFlOTcyZGViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ig-H_Q-XP71Pc0ekriTW3hwvzBSGZqyjLj3Q_UAGYWY)

yep maybe is a permissions stufs missed need look

**Comment by Deswa16** [comment]
I understood from what you wrote that you edited it, but I just wanted to ask a question, do you have a chance to do this?

**Comment by Furtif** [comment]
> I understood from what you wrote that you edited it, but I just wanted to ask a question, do you have a chance to do this?

there will certainly be, waiting for it to be resolved, you will be able to use 1.85 https://github.com/cedricp/ddt4all/wiki/Android-port

**Comment by Furtif** [comment]
I don't think version 1.85 is on this list!!!

[EcuTweaker-release-v1.85.apk.zip](https://github.com/cedricp/ecutweaker/files/15143479/EcuTweaker-release-v1.85.apk.zip)

**Comment by Bobbik1** [comment]
1.85 manages BT connections well, any newer version not connects to bluetooth device at all.

**Comment by davidwkerr** [comment]
I had the same problem with version 2.1 on Pixel 7A and Android 14. When I click on select and see paired devices, click on the Konnwei KW902 or OBDLink LX and there is a very quick message "no devices found" and the status window shows bluetooth connection failed.

But now, 1.85 is working. But I am not sure how to enable writing parameters with it. Yes, I have made a donation.....

**Comment by igogo2** [comment]
The same story on LG G8x Android 10.
Version 2.1 release and debug version can't connect to BW ELM. No action. Storage and location permitions are enabled.
Version 1.81 works well.
Unfortunately i do not have posiibility to use logcat now.

**Comment by Furtif** [comment]
@manio you have test last artifatc https://github.com/cedricp/ecutweaker/actions/runs/11785612961 ??? i have try in android 13 usb works, wifi not, but i not have a bluetooh for test

**Comment by manio** [comment]
@Furtif
I just installed your debug version from mentioned artifact on the same Samsung phone (Android 11) and it doesn't help. Clicking on button = no reaction. Old version 1.8 still working fine.

**Comment by Furtif** [comment]
> @Furtif I just installed your debug version from mentioned artifact on the same Samsung phone (Android 11) and it doesn't help. Clicking on button = no reaction. Old version 1.8 still working fine.

ok thanks

**Comment by manio** [comment]
@haiku255 And that's it? :)
@Furtif can you prepare an apk with this change?

**Comment by manio** [comment]
@haiku255 If you don't require permission it won't work because I had to uninstall previous version before installing yours (so the granted permission will go away), besides this issue is about Android 11, not 14.

**Comment by manio** [comment]
Tested your version - same problem on my phone. Clicking "Select device" for bluetooth => no action.

**Comment by Furtif** [comment]
> @manio
> Here is my apk if you want to test it now, be careful all the required permissions are not requested at the first start of the application, has been tested on a Galaxy S9 with Android 14.
> https://drive.proton.me/urls/0J536C04C0#opd8CcKXlCQC

Pr change if you want please.

**Comment by Furtif** [comment]
But I thinks have same issue in recent android, I have changed to this because warn in sdk recent

**Comment by Furtif** [comment]
> I added a location permission request for older androids, on the android studio emulator I used a phone with android 11, which had the error where nothing happened after the click, but now it seems to work. On my emulator the storage permission is not displayed and I did not try to update the requests so make sure to authorize it manually. Here is the apk to see if everything works correctly https://drive.proton.me/urls/E8WPQS7JFW#HkM8jJEipe2o

### PR you changes
 - or propose

**Comment by Furtif** [comment]
> Sorry I have never used github except to download codes, I don't know how it works.
> 
> Here is the modified code in MainActivity.java
> 
> ```
> void selectDevice(){
>         BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
> 
>         // If the adapter is null, then Bluetooth is not supported
>         if (bluetoothAdapter == null) {
>             return;
>         }
> 
>         if (bluetoothAdapter.isEnabled()) {
>             stopConnectionTimer();
>             try {
>                 if (mLinkMode == LINK_BLUETOOTH) {
>                     // ajout perso \/
>                     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
>                         // Android 12 (31) +
>                         if (!askBluetoothScanPermission()) {
>                             return;
>                         }
>                     } else {
>                         // Android 11 (30) -
>                         if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
>                             ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, REQUEST_PERMISSION_LOCATION);
>                             return;
>                         }
>                     }
>                     // ajout perso /\
>                     
>                     Intent serverIntent = new Intent(this, DeviceListActivity.class);
>                     startActivityForResult(serverIntent, REQUEST_CONNECT_DEVICE);
>                 } else if (mLinkMode == LINK_USB) {
>                     Intent serverIntent = new Intent(this, UsbDeviceActivity.class);
>                     startActivityForResult(serverIntent, REQUEST_CONNECT_DEVICE);
>                 }
>             } catch (android.content.ActivityNotFoundException e) {
> 
>             }
>         }
>     }
> ```

ok send me you change in discord https://discord.gg/HPvvSa8J

**Comment by Furtif** [comment]
send me you files changed i go change it at a pr for test

**Comment by Furtif** [comment]
@manio https://github.com/cedricp/ecutweaker/actions/runs/12188586212 thanx to @haiku255  , you test it please ?

**Comment by Furtif** [comment]
@haiku255 thank for https://github.com/cedricp/ecutweaker/pull/14 , now need test for look

**Comment by manio** [comment]
@Furtif I tested the last artifact (release):
Yeah!!! It's finally working back!!!!!
For the first time it was asking for permission, second click is giving me a list of Bluetooth devices!

Thank you @haiku255 for your work!

**Comment by Furtif** [comment]
> @Furtif I tested the last artifact (release): Yeah!!! It's finally working back!!!!! For the first time it was asking for permission, second click is giving me a list of Bluetooth devices!
> 
> Thank you @haiku255 for your work!

is fine ??? but in pixel5 a14 not work !

**Comment by Furtif** [comment]
but ok go merge if ok and close this

**Comment by Furtif** [comment]
@manio test new release https://github.com/cedricp/ecutweaker/releases/tag/v2.2

**Comment by manio** [comment]
@Furtif 
Confirm, working OK! :)

Thank you all involved!

**Reply by Furtif** [reply]
Click `Mark as answer`

**Reply by Furtif** [reply]
@manio join https://discord.gg/cBqDh9bTHP

**Comment by Furtif** [comment]
> @Furtif Confirm, working OK! :)
> 
> Thank you all involved!

cool read and test https://github.com/cedricp/ecutweaker/pull/18

**Comment by igogo2** [comment]
Unfortunately for me doesn't fixed. Selecting work but connecting not.
I've tested four versions 2.2 and 2.3 (release and debug) on two phones (Android 9 and 10).
After clicking "Select device" button with Bluetooth "Paired devices" window opened correctly and showed my ELM. After clicking on it the window closes but pictogram still show "unconnected" and "Auto identify" doesn't work.
How could i provide any additional (debug) information for identifying the problem?

> ![Auto identify](https://github.com/user-attachments/assets/491813ba-b7bc-444a-b690-a22bc1ccfd51)

> ![Read codes](https://github.com/user-attachments/assets/ab7d3c61-840d-4253-a023-c196b902e391)

**Reply by Furtif** [reply]
about you pic try rebuid apk in androidStudio at 
this line https://github.com/cedricp/ecutweaker/blob/master/EcuTweaker/src/main/java/org/quark/dr/canapp/MainActivity.java#L321 remove the `=` and test if ok please.

**Reply by Furtif** [reply]
` if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) ` to ` if (Build.VERSION.SDK_INT > Build.VERSION_CODES.S)`

**Reply by Furtif** [reply]
if not ok issue is in this https://github.com/cedricp/ecutweaker/blob/master/EcuTweaker/src/main/java/org/quark/dr/canapp/MainActivity.java#L704 i need maybe make that (all in one android 11 > up) or bypass id < 11...

**Reply by Furtif** [reply]
hum go look if i make a better conditional .

**Reply by Furtif** [reply]
but ODDD you have acepted LOCATION permition ? in A9-10 bluethood is Croate_Location you have Acepted ? IF not sure please uninstall app reinstall and accordings premitions.

**Reply by igogo2** [reply]
I accepted all 2 permissions Location and Storage Before first run of application.

**Comment by manio** [comment]
@igogo2 
Oooops... I didn't check that... I was only focusing on "device selection" window which was not working and eventually has been fixed.
Ok... but is it working fine for you in "known-good" version 1.85?

**Reply by igogo2** [reply]
Yes. I still been using 1.85 version. It works on my phones.

**Reply by Furtif** [reply]
🤔🤔🧐

**Reply by Furtif** [reply]
@igogo2 try last artifact https://github.com/cedricp/ecutweaker/actions/runs/12569822757/artifacts/2376139502

**Reply by Furtif** [reply]
@manio issues ?? for my side wifi + usb all right

**Reply by Furtif** [reply]
🤷🏻‍♂️

**Reply by manio** [reply]
@Furtif 
Unfortunately @igogo2 is right! It is not connecting also for me. As I said i was focusing to test the selection dialog but I didn't testing it on real device. Now I tested some recent versions and:
1.85 - still working OK (showing dialog and connecting)
2.3 and your last artifact - showing bluetooth device selection dialog but then it is not connecting :(

I also can see that old version when I connected for the first time asked if I want to Pair. New versions seems is doing nothing - not asking and not doing anything - similar as was with "Select device" button.

The problem is visible like this:
old version: clicking on red "⛓️‍💥" symbol: asking for pairing, connect and the icon is green (connected)
new versions: clicking on red "⛓️‍💥" symbol is doing nothing = no action (like it was with "select device" before)

All tests done on Android 11 (Samsung Galaxy A10)

**Reply by Furtif** [reply]
🤷🏻‍♂️🤔🤔🧐 odd, Clicking on the icon normally doesn't work, I think this is just the connection status, as I said in discord.

**Reply by igogo2** [reply]
> @igogo2 try last artifact https://github.com/cedricp/ecutweaker/actions/runs/12569822757/artifacts/2376139502
Doesn't work.

**Comment by manio** [comment, marked as answer]
Tested https://github.com/cedricp/ecutweaker/pull/30 and this fixed the problem for me on Android 11 (Samsung Galaxy A10)
Thank you @Furtif

**Reply by Furtif** [reply]
Cool 👌🥳
